File: sort.py
#!/usr/bin/python
# -*- coding: utf-8 -*-


# bubblesort
# 大循环： 从第一个元素开始，到最后一个元素
# 小循环： 对所有该小循环中的元素，两两比较元素大小，如果有根据大小，交换两个元素的位置

def bubbleSort(arr):
	for i in range (0,len(arr)):
		for j in range(i+1,len(arr)):
			if arr[i]>arr[j]:
				temp=arr[i]
				arr[i]=arr[j]
				arr[j]=temp
	return arr



# selectionsort
# 从arr[0]~arr[len(arr)-1]中找出最小的元素及它的的标志位min，将arr[min]与arr[0]交换
# 从arr[1]~arr[len(arr)-1]中找出最小的元素及它的的标志位min，将arr[min]与arr[1]交换

def selectionSort(arr):
	for i in range (0,len(arr)):
		min = i
		for j in range (i+1,len(arr)):
			if arr[min]>arr[j]:
				min=j
		arr[i]=arr[min]
		arr[j]=arr[i]
	return arr



# insertsort
# 对整个序列，从前往后扫描  取待排序序列。 把最后一个元素作为寻找插入位的元素（待排序元素）；
# 对取出的待排序序列，从后往前扫描  寻找插入位置。 当key更小时往前插一位()

def insertionSort(arr):
	for i in range(1,len(arr)):
		j=i-1
		key=arr[i]
		while j>=0:
			if key<arr[j]:
				a[j+1]=a[j]
				j-=1
			arr[j+1]=key
	return arr



# shellsort
# 判断len(list)/3 > (gap=1),才可以用shellsort。 满足时，gap=3*gap+1
# 取 i range in（gap,len(arr)）, 比较 arr[i] a[i-gap]
# 每次 gap=gap/3， 直到 gap=math.floor(gap/3)=0

import math
def shellSort(arr):
	import math
	gap=1
	while(gap<len(arr)/3):
		gap=gap*3+1              # set gap value
	while(gap>0):
		for i in range(gap,len(arr)):       # range(gap,len(arr)) use insertsort
			temp=arr[i]
			j=i-gap
			while j>=o and arr[j]>temp:
				arr[j+gap]=arr[j]
				j-=gap
			arr[j+gap]=temp                
		gap=math.floor(gap/3)            # reflesh gap value
	return arr



# merge sort
# 将待排序序列拆分成更小的子序列
# 序列拆分直到子序列仅剩两个元素 可进行比较并排序，对已排好的两个序列里的元素比较，将取出的较大（或较小）的元素放入一个新的空序列中

def merge(left,right):
	temp=[]
	i=j=0
	while i<len(left) and j<len(right):
		if left[i]<right[j]:
			temp.append(left[i])
			i+=1
		else:
			temp.append(right[j])
			j+=1
		if i=len(left):
			for x in right[j:]:
				temp.append(x)
		else if j=len(right):
			for x in left[i:]:
				temp.append(x)
		return temp

def mergeSort(arr):
	if len(arr)<2:
		return arr
	middle=len(arr)/2
	left=mergeSort(arr[:middle])
	right=mergeSort(arr[middle:])
	return merge(left,right)



# quicksort 
# 取出基准元素，比它小的元素放 low[],大的放high[], 并拿到基准元素的标志位key_index（arr[key_index]=key）
# 分别对low[]， high[] 进行快速排序

def subSort(array,low,high):
    key = array[low]
    while low < high:
        while low < high and array[high] >= key:
            high -= 1
        while low < high and array[high] < key:
            array[low] = array[high]
            low += 1
            array[high] = array[low]
    array[low] = key
    return low

def quickSort(array,low,high):
     if low < high:
        key_index = subSort(array,low,high)
        quickSort(array,low,key_index)
        quickSort(array,key_index+1,high)



# heapsort
# 直接选择排序的改进版(选出合适的元素索引，再做一次交换)。
# len(A)=n, A[i]出节点： 双亲节点为 A[(i-1)/2]
# 			  		  左孩子节点为 A[2i+1]， 若2i+1>n时，A[i]无左孩子节点，为叶子
# 			  		  右孩子节点为 A[2i+2]， 若2i+2>n时，A[i]无右孩子节点
# 从最后一个非叶子节点A[i]开始（即A[n/2-1]），将以A[i]为根节点的二叉树调整为堆 （目标：所有的子树都为堆的完全二叉树）
# 1、建堆：大根堆（从A[n/2-1]开始，目标是A[0]最大）；2、调堆：A[0]与A[lenth-1]互换位置，输出末尾节点，再对(n-2)的树调整，重新组成大根堆

def adjustHeap(arr, i, arrLen):
	max=i
	left=2*i+1
	right=2*i+2
	if i<arrLen/2:
		if left< arrLen and arr[left]>arr[max]:
			max=left
		if right<arrLen and arr[right]>arr[max]:
			max=right
		if i!=max:
			arr[i]=arr[max]
			adjustHeap(arr, max, arrLen)

def buildHeap(arr,arrLen):
	for i in range (0,arrLen/2)[::-1]:
		adjustHeap(arr, i, arrLen)

def heapSort(arr):
	arrLen=len(arr)
	buildHeap(arr,arrLen)
	for i in range (0,arrLen)[::-1]:
		arr[0],arr[i]=arr[i],arr[0]
		adjustHeap(arr, 0, i)

    # array = [1, 2, 5, 3, 6, 8, 4]
	# >>> array[::2]
	# [1, 5, 6, 4]
	# >>> array[2::]
	# [5, 3, 6, 8, 4]
	# >>> array[::3]
	# [1, 3, 4]
	# >>> array[::4]
	# [1, 6] 
	# 如果想让他们颠倒形成reverse函数的效果
	# >>> array[::-1]
	# [4, 8, 6, 3, 5, 2, 1]
	# >>> array[::-2]
	# [4, 6, 5, 1]

if __name__ == '__main__':
    arr = rawInput()
    # sort(arr)
    print mergSort(arr)

File: sort.cpp
/*  feature:   直接选择排序的改进版
	编号为i的节点：
	    1\ 双亲编号为 i/2
	    2、若2i<n, ki左孩子节点为2i； 否则ki为叶子
	    3、若2i+1<=n，ki的右孩子节点为2i+1； 否则ki无右孩子节点

	从最后一个非叶子节点i开始，将以i为根节点的二叉树调整为堆 （目标：所有的子树都为堆的完全二叉树）
 */


/*
    输入：数组A，堆的长度hLen，以及需要调整的节点i
    功能：调堆
 */
void AdjustHeap(int A[], int hLen, int i){
	// int left=LeftChild(i);
	// int right=RightChild(i);
	int largest=i;
	int temp;

	while(left<hLen || right<hLen){
		if(left<hLen && A[largest]<A[left]){
			largest=left;
		}
		if(right<hLen && A[largest]<A[right]){
			largest=right;
		}

		if(i!=largest){              //father-node is not the max
			temp=A[largest];
			A[largest]=A[i];
			A[i]=temp;

			i=largest;
			left=LeftChild(i);
			right=RightChild(i);
		}
		else{
			break;
		}
	}
}

void BuildHeap(int A[], int aLen){
	int i;
	int begin=hLen/2-1;        // last non-leafnode
	for(i=begin; i>=0; i--){
		AdjustHeap(A, hLen, i);
	}
}

void Heapsort(int A[], int aLen){
	int hLen=aLen;
	int temp;

	BuildHeap(A, hLen);        // set heap

	while(hLen>1){
		temp=A[hLen-1];       // exchange A[0] and A[hLen]
		A[hLen-1]=A[0];
		A[0]=temp;
		hLen--;            // size of heap -1
		AdjustHeap(A,hLen,0);
	}
}





// package sort;
 
// import java.util.Arrays;
// import util.MathUtil;
// /**
//  * 通过大顶堆实现堆排序，升序排序
//  * 
//  */
 
// public class HeapSort {
//     public static void main(String[] args) {
//         int[] arr={9,6,12,32,23,11,2,100,85};
//         sort(arr);
//         System.out.println(Arrays.toString(arr));
//     }
//     //这里将i定义为完全二叉树的根
//     //将完全二叉树调整为大顶堆,前提是二叉树的根的子树已经为大顶堆。
//     public static void adjustHeap(int[]a ,int i,int size){
//         int lChild=2*i+1;        //左孩子
//         int rChild=2*i+2;        //又孩子
//         int max=i;                //临时变量
//         if(i<size/2){            //如果i是叶子节点就结束
//             if(lChild<size&&a[max]<a[lChild])
//                 max=lChild;
//             if(rChild<size&&a[max]<a[rChild])
//                 max=rChild;
//             if(max!=i){
//                 MathUtil.swap(a, max, i);//交换后破环了子树的堆结构
//                 adjustHeap(a, max, size);//递归，调节子树为堆
//             }
//         }
//     }
     
//     //建立堆，堆是从下往上建立的，因为adjustHeap函数是建立在子树已经为大顶堆。
//     public static void buildHeap(int[]a,int size){    
//         for(int i=size/2;i>=0;i--){//从最后一个非叶子节点，才能构成adjustHeap操作的目标二叉树
//             adjustHeap(a, i, size);
//         }        
//     }
 
//     //将数组分为两部分，一部分为有序区，在数组末尾，另一部分为无序区。堆属于无序区
//     public static void sort(int[] arr){
//         int size=arr.length;
//         buildHeap(arr, size);
//         for(int i=size-1;i>0;i--){//i为无序区的长度，经过如下两步，长度递减
//             //堆顶即下标为0的元素
//             MathUtil.swap(arr, i, 0);//1.每次将堆顶元素和无序区最后一个元素交换，即将无序区最大的元素放入有序区
//             adjustHeap(arr, 0, i);   //2.将无顺区调整为大顶堆，即选择出最大的元素。
//         }
//     }    
// }


